src/uct/tcp/tcp_ep.c

```diff
@@ -217,13 +220,147 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_ep_t, uct_tcp_iface_t *iface,
+                                    UCS_BIT(UCT_TCP_EP_CTX_TYPE_TX))) {
+            uct_tcp_cm_remove_ep(iface, ep);
+        }
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_ep_remove_ctx_cap(uct_tcp_ep_t *ep,
+                                       uct_tcp_ep_ctx_type_t cap)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uint8_t prev_caps      = ep->ctx_caps;
+
+    ep->ctx_caps &= ~UCS_BIT(cap);
+    uct_tcp_ep_trace_ctx_caps_change(ep, prev_caps);
+
+    if (!uct_tcp_ep_peer_addr_to_itself(ep)) {
+        if (prev_caps == (UCS_BIT(UCT_TCP_EP_CTX_TYPE_RX) |
+                          UCS_BIT(UCT_TCP_EP_CTX_TYPE_TX))) {
+            return uct_tcp_cm_add_ep(iface, ep);
+        } else if (!ep->ctx_caps) {
+            uct_tcp_cm_remove_ep(iface, ep);
+        }
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t uct_tcp_ep_move_ctx_cap(uct_tcp_ep_t *from_ep, uct_tcp_ep_t *to_ep,
+                                     uct_tcp_ep_ctx_type_t ctx_cap)
+{
+    ucs_status_t status;
+
+    status = uct_tcp_ep_remove_ctx_cap(from_ep, ctx_cap);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return uct_tcp_ep_add_ctx_cap(to_ep, ctx_cap);
+}
+
+static void uct_tcp_ep_dup(uct_tcp_iface_t *iface, uct_tcp_ep_t *from_ep)
+{
+    uct_tcp_ep_t *ep;
+    ucs_status_t status;
+    uint32_t events;
+
+    /* Duplicate this EP and then close the current one */
+    status = uct_tcp_ep_init(iface, from_ep->fd, &from_ep->peer_addr, &ep);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = uct_tcp_ep_move_ctx_cap(from_ep, ep,
+                                     UCT_TCP_EP_CTX_TYPE_RX);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    uct_tcp_cm_set_conn_state(ep, from_ep->conn_state);
+    /* we do it here to avoid calling `uct_tcp_cm_change_conn_state`
+     * in the cleanup procedure */
+    uct_tcp_cm_set_conn_state(from_ep, UCT_TCP_EP_CONN_STATE_CLOSED);
+
+    ep->rx              = from_ep->rx;
+    memset(&from_ep->rx, 0, sizeof(from_ep->rx));
+
+    events              = from_ep->events & EPOLLIN;
+    uct_tcp_ep_mod_events(from_ep, 0, from_ep->events);
+    from_ep->fd         = -1;
+    if (events) {
+        uct_tcp_ep_mod_events(ep, EPOLLIN, 0);
+    }
+
+    ucs_debug("tcp_ep %p: created duplicate EP for %p RX events "
+              "on iface %p, fd %d", ep, from_ep, iface, ep->fd);
+    return;
+
+ err:
+    ucs_error("tcp_ep %p: unable to create copy for RX context", from_ep);
+}
+
 static UCS_CLASS_CLEANUP_FUNC(uct_tcp_ep_t)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(self->super.super.iface,
                                             uct_tcp_iface_t);
 
+    if ((self->ctx_caps == (UCS_BIT(UCT_TCP_EP_CTX_TYPE_RX) |
+                            UCS_BIT(UCT_TCP_EP_CTX_TYPE_TX))) &&
+        (self->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED)) {
+        uct_tcp_ep_dup(iface, self);
```

Comment:
better use proxy func

Reply:
removed and implemented a proxy func:
1. `uct_tcp_ep_destroy` - the proxy func that is called from `uct_ep_destroy`
2. `uct_tcp_ep_destroy_internal` - the function with dtor

Reply:
why is it still used?

Reply:
I forgot to remove `uct_tcp_ep_dup` and its usage - now fixed